crates/kubelet/src/state.rs

```diff
@@ -206,26 +216,83 @@ impl<PodState> Transition<PodState> {
     /// }
     /// ```
     #[allow(clippy::boxed_local)]
-    pub fn next<I: State<PodState>, S: State<PodState>>(_i: Box<I>, s: S) -> Transition<PodState>
+    pub fn next<I: State<ProviderState, PodState>, S: State<ProviderState, PodState>>(
+        _i: Box<I>,
+        s: S,
+    ) -> Transition<ProviderState, PodState>
     where
         I: TransitionTo<S>,
     {
         Transition::Next(StateHolder { state: Box::new(s) })
     }
+
+    /// Represents a transition to a new state that is not checked against the
+    /// set of permissible transitions. This is intended only for use by generic
+    /// states which cannot declare an exit transition to an associated state
+    /// without encountering a "conflicting implementations" compiler error.
+    #[allow(clippy::boxed_local)]
+    pub fn next_unchecked<I: State<ProviderState, PodState>, S: State<ProviderState, PodState>>(
+        _i: Box<I>,
+        s: S,
+    ) -> Transition<ProviderState, PodState> {
+        Transition::Next(StateHolder { state: Box::new(s) })
+    }
 }
 
+// TODO: consider moving this to the ProviderState
 #[async_trait::async_trait]
 /// Allow for asynchronous cleanup up of PodState.
 pub trait AsyncDrop: Sized {
+    /// The type of any provider-level state from which the pod
+    /// needs to be cleaned up.
+    type ProviderState;
     /// Clean up PodState.
-    async fn async_drop(self);
+    async fn async_drop(self, provider_state: &mut Self::ProviderState);
+}
+
+/// Provides shared access to provider-level state between multiple pod
+/// state machines running within the provider.
+pub struct SharedState<T> {
+    state: std::sync::Arc<tokio::sync::RwLock<T>>,
+}
+
+impl<T> Clone for SharedState<T> {
+    fn clone(&self) -> Self {
+        Self {
+            state: self.state.clone(),
+        }
+    }
+}
+
+impl<T> SharedState<T> {
+    /// Creates a SharedState to provide shared access to the specified value.
+    pub fn new(value: T) -> Self {
+        Self {
+            state: std::sync::Arc::<_>::new(tokio::sync::RwLock::new(value)),
+        }
+    }
+
+    /// Acquires a read lock for the shared state.
+    pub async fn read(&self) -> tokio::sync::RwLockReadGuard<'_, T> {
+        self.state.read().await
+    }
+
+    /// Acquires a write lock for the shared state.
+    pub async fn write(&self) -> tokio::sync::RwLockWriteGuard<'_, T> {
+        self.state.write().await
+    }
 }
 
 #[async_trait::async_trait]
 /// A trait representing a node in the state graph.
-pub trait State<PodState>: Sync + Send + 'static + std::fmt::Debug {
+pub trait State<ProviderState, PodState>: Sync + Send + 'static + std::fmt::Debug {
     /// Provider supplies method to be executed when in this state.
-    async fn next(self: Box<Self>, pod_state: &mut PodState, pod: &Pod) -> Transition<PodState>;
+    async fn next(
+        self: Box<Self>,
+        provider_state: SharedState<ProviderState>,
```

Comment:
I still believe that adding this additional state adds enormous complexity to the API (as can be seen in this pull request) without enabling any new functionality.

Reply:
That is fair.  It doesn't add new functionality but it provides a way of _talking about_ the state of the provider rather than tucking it away via a reference in the pod state accessible only in a provider-specific way.  We could have it so that `GenericPodState` defines a `provider_state()` method, and states that are interested in the shared state get it that way.  I don't have a strong opinion either way, but we do need _some_ explicit way to talk about shared state.

I'm not sure 'enormous complexity' is fair - this change definitely has a ripple effect which is why the PR is so big, but I'm not convinced breaking out the shared state from the pod state truly adds complexity.  But that is just my perception and is skewed by the way I think about the way the providers manage state...

Reply:
I think the complexity im referring to is that now everything has two generic type parameters. Which results in a lot of verbosity in code and in error messages.

While I appreciate being able to talk about the states in this way, I think it is wrong for these types to be generic in two dimensions when there will nearly always be a 1:1 relationship between the two types in any given implementation. 

One alternative that I could suggest is to spend some time on documentation in the Provider module to specifically call out the shared state pattern and how it should be implemented, caveats (don’t hold locks for a long time), etc. I think this will be useful whether we go with the pattern in this branch or not. I’d be happy to take a crack at this.